src/kOS/Suffixed/Part/PartValueFactory.cs

```diff
@@ -20,21 +20,28 @@ public static ListValue<PartValue> ConstructGeneric(IEnumerable<global::Part> pa
 
         public static PartValue Construct(global::Part part, SharedObjects shared)
         {
-            var multiModeEngines = part.Modules.GetModules<MultiModeEngine>();
-            if (multiModeEngines.Count > 0)
-                return new EngineValue(part, multiModeEngines.First(), shared);
-
-            var moduleEngines = part.Modules.GetModules<ModuleEngines>();
-            if (moduleEngines.Count > 0)
-                return new EngineValue(part, new ModuleEngineAdapter(moduleEngines.First()), shared);
-
-            var moduleDockingNodes = part.Modules.GetModules<ModuleDockingNode>();
-            if (moduleDockingNodes.Count > 0)
-                return new DockingPortValue(moduleDockingNodes.First(), shared);
-
-            var moduleEnviroSensors = part.Modules.GetModules<ModuleEnviroSensor>();
-            if (moduleEnviroSensors.Count > 0)
-                return new SensorValue(part, moduleEnviroSensors.First(), shared);
+            PartValue ret = null;
+            for (int i = 0; i < part.Modules.Count; ++i)
+            {
+                if (part.Modules[i] is MultiModeEngine)
+                {
+                    return new EngineValue(part, (MultiModeEngine)part.Modules[i], shared);
+                }
+                if (part.Modules[i] is ModuleEngines)
+                {
+                    ret = new EngineValue(part, new ModuleEngineAdapter((ModuleEngines)part.Modules[i]), shared);
+                }
+                else if (part.Modules[i] is ModuleDockingNode)
+                {
+                    ret = new DockingPortValue((ModuleDockingNode)part.Modules[i], shared);
+                }
+                else if (part.Modules[i] is ModuleEnviroSensor)
+                {
+                    ret = new SensorValue(part, (ModuleEnviroSensor)part.Modules[i], shared);
+                }
+            }
```

Comment:
While there is no searchable index of `part.Modules`, the `PartModule` class does support a `name` property which is a `string` and should be the name of the type.  I didn't do any testing to see if type checking or string comparison is faster.  But it may be worth while.

I would prefer using LINQ just because they have some clever optimizations (partial ordering of lists to minimize comparisons when you find a match part way through, that kind of thing).  But it is allocation heavy, and the Mono GC that Unity uses doesn't handle it very well.